### PR TITLE
Add first version of PyTorch notebook

### LINKED ISSUE #40: Pytorch notebook
- Show what is tensors
- Show autograd
- Show sampling using no gradients and CUQIpy
- Show CUQIpy using NUTS when possible (and show when it cant)
- Show CUQIpy-PyTorch to automatically do gradients.

### PR DESCRIPTION
**Pull request by nabriis**
Closes #40

### CODE REVIEW

File: training/CUQIpy-PyTorch/HMC.ipynb [hunk 1]
```diff
@@ -0,0 +1,857 @@
...
+    "        grad = donut.gradient(posg)\n",
+    "        norm = np.linalg.norm(grad, axis=0)\n",
+    "        u, v = grad[0, :]/norm, grad[1, :]/norm\n",
+    "\n",
+    "    # plot PDF and gradient\n",
+    "    plt.contourf(X, Y, Z, nl)\n",
+    "    plt.contour(X, Y, Z, nl, linewidths=0.5, colors='k') \n",
+    "    if plot_grad:\n",
+    "        plt.quiver(posg[:, 0], posg[:, 1], u, v, units='xy', scale=4, color='gray')\n",
+    "\n",
```
Comment (jakobsj) [inline training/CUQIpy-PyTorch/HMC.ipynb:119]: Split cell here to separate function def and the call to generate the plot. Add a new markdown cell in between to simply say we now plot using the function defined.

File: training/CUQIpy-PyTorch/HMC.ipynb [hunk 2]
```diff
@@ -0,0 +1,857 @@
...
+    "We can then compare the samples obtained from the two methods by plotting the samples in a \"pair plot\" (a scatter plot of each pair of samples)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "samples_MH.plot_pair()\n",
```
Comment (jakobsj) [inline training/CUQIpy-PyTorch/HMC.ipynb:162]: add `plt.axis('equal') after this line

File: training/CUQIpy-PyTorch/HMC.ipynb [hunk 3]
```diff
@@ -0,0 +1,857 @@
...
+    "plt.title('Metropolis-Hastings')\n",
+    "samples_NUTS.plot_pair()\n",
+    "plt.title('Hamiltonian Monte Carlo (NUTS)');"
```
Comment (jakobsj) [inline training/CUQIpy-PyTorch/HMC.ipynb:165]: and after this

File: training/CUQIpy-PyTorch/HMC.ipynb [hunk 4]
```diff
@@ -0,0 +1,857 @@
...
+    "**The examples in here are for illustration purposes only. You do not need to understand the all details of automatic differentiation to use CUQIpy-PyTorch.**"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In the previous section we used the NUTS implementation from the main CUQIpy package, which requires the gradient of the log-posterior to be manually defined in the `Distribution` class.\n",
+    "\n",
+    "In this section we see hwo to use PyTorch tensors and automatic differentiation to compute the gradient of any distribution. The obvious benefit here is that we do not need to manually define the gradient of the logpdf, which makes the modelling much more flexible.\n",
```
Comment (jakobsj) [inline training/CUQIpy-PyTorch/HMC.ipynb:229]:
```suggestion
    "In this section we see how to use PyTorch tensors and automatic differentiation to compute the gradient of any distribution. The obvious benefit here is that we do not need to manually define the gradient of the logpdf, which makes the modelling much more flexible.\n",
```

File: training/CUQIpy-PyTorch/HMC.ipynb [hunk 5]
```diff
@@ -0,0 +1,857 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Really utilizing the power of automatic differentiation\n",
+    "\n",
+    "One of the main use-cases for automatic differentiation in the context of Bayesian inference is the compute the gradient of more complex Bayesian models. To illustrate this, suppose one part of the Bayesian model contains a Gaussian distribution with a covariance matrix that is a function of another random variable, say $s$. The Gaussian could be written as follows.\n",
+    "\n",
+    "$$\n",
```
Comment (jakobsj) [inline training/CUQIpy-PyTorch/HMC.ipynb:419]: This math is unrendered on colab
image

File: training/CUQIpy-PyTorch/HMC.ipynb [hunk 6]
```diff
@@ -0,0 +1,857 @@
...
+    "Instead, we use thin wrappers around PyTorch distributions, that acts as a drop-in replacement for the `cuqi` distributions. They are imported as follows."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from cuqipy_pytorch.distribution import Gaussian, HalfGaussian, LogGaussian, Uniform, Gamma, StackedJointDistribution"
```
Comment (jakobsj) [inline training/CUQIpy-PyTorch/HMC.ipynb:359]: I'm starting to round on cuqi_torch...   ;-)

**Review by jakobsj — APPROVED** [review]
Brilliant notebook, the power is clear!  Perhaps a tad long but there is so much to say - one might consider splitting in several more focused notebooks, but that's for another time. Just a few tiny fixes needed.

File: training/CUQIpy-PyTorch/HMC.ipynb (hunk 1 above)
Comment (nabriis) [reply training/CUQIpy-PyTorch/HMC.ipynb:119]: Good suggestion. Thanks!

File: training/CUQIpy-PyTorch/HMC.ipynb (hunk 2 above)
Comment (nabriis) [reply training/CUQIpy-PyTorch/HMC.ipynb:162]: Thanks!

File: training/CUQIpy-PyTorch/HMC.ipynb (hunk 5 above)
Comment (nabriis) [reply training/CUQIpy-PyTorch/HMC.ipynb:419]: Thanks
